### PR TITLE
LUCENE-9378: Make it possible to configure how to trade speed for compression on doc values.

### PR DESCRIPTION
**Pull request by jpountz**
This adds a switch to `Lucene80DocValuesFormat` which allows to
configure whether to prioritize retrieval speed over compression ratio
or the other way around. When prioritizing retrieval speed, binary doc
values are written using the exact same format as before more aggressive
compression got introduced.

### CODE REVIEW

File: lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/lucene87/Lucene87Codec.java [hunk 1]
```diff
@@ -56,6 +57,23 @@
...
+  public static enum Mode {
+    /** Trade compression ratio for retrieval speed. */
+    BEST_SPEED(Lucene87StoredFieldsFormat.Mode.BEST_SPEED, Lucene80DocValuesFormat.Mode.BEST_SPEED),
+    /** Trade retrieval speed for compression ratio. */
+    BEST_COMPRESSION(Lucene87StoredFieldsFormat.Mode.BEST_COMPRESSION, Lucene80DocValuesFormat.Mode.BEST_COMPRESSION);
+
+    private final Lucene87StoredFieldsFormat.Mode storedMode;
+    private final Lucene80DocValuesFormat.Mode dvMode;
+
+    private Mode(Lucene87StoredFieldsFormat.Mode storedMode, Lucene80DocValuesFormat.Mode dvMode) {
```
Comment (mikemccand) [inline lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/lucene87/Lucene87Codec.java:71]: Nice!  So we roll up the tradeoffs to Codec level which will then tell each format how to tradeoff.

File: lucene/core/src/test/org/apache/lucene/codecs/lucene80/BaseLucene80DocValuesFormatTestCase.java [hunk 2]
```diff
@@ -286,7 +278,7 @@ private void doTestTermsEnumRandom(int numDocs, Supplier<String> valuesProducer)
     conf.setMergeScheduler(new SerialMergeScheduler());
     // set to duel against a codec which has ordinals:
     final PostingsFormat pf = TestUtil.getPostingsFormatWithOrds(random());
-    final DocValuesFormat dv = new Lucene80DocValuesFormat();
+    final DocValuesFormat dv = getCodec().docValuesFormat();
```
Comment (mikemccand) [inline lucene/core/src/test/org/apache/lucene/codecs/lucene80/BaseLucene80DocValuesFormatTestCase.java:281]: Will this randomize between the different `Mode` tradeoffs?

File: lucene/core/src/test/org/apache/lucene/codecs/lucene80/TestBestSpeedLucene80DocValuesFormat.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ */
+package org.apache.lucene.codecs.lucene80;
+
+import org.apache.lucene.codecs.Codec;
+import org.apache.lucene.util.TestUtil;
+
+/**
+ * Tests Lucene80DocValuesFormat
+ */
+public class TestBestSpeedLucene80DocValuesFormat extends BaseLucene80DocValuesFormatTestCase {
```
Comment (mikemccand) [inline lucene/core/src/test/org/apache/lucene/codecs/lucene80/TestBestSpeedLucene80DocValuesFormat.java:25]: Do we also have a dedicated `TestBestCompressedLucene80DocValuesFormat`?

**Review by mikemccand — COMMENTED** [review]
Thank you @jpountz!  I left a few small questions, but this looks great.

At write time, Lucene user can choose whether they want smaller index (compressed) or faster search (`BEST_SPEED`), and the resulting index has Lucene's normal full back compat guarantee.

Let's make sure Lucene's `test-framework` is fully randomize these write-time `Codec` choices so we get good test coverage of all options.

File: lucene/core/src/test/org/apache/lucene/codecs/lucene80/TestBestSpeedLucene80DocValuesFormat.java (hunk 3 above)
Comment (mikemccand) [reply lucene/core/src/test/org/apache/lucene/codecs/lucene80/TestBestSpeedLucene80DocValuesFormat.java:25]: Oh nevermind I see you opened followon issue for this: https://issues.apache.org/jira/browse/LUCENE-9602

Comment (jpountz) [reply lucene/core/src/test/org/apache/lucene/codecs/lucene80/TestBestSpeedLucene80DocValuesFormat.java:25]: You should see a `TestBestCompressedLucene80DocValuesFormat` file as well in this PR. I opened LUCENE-9602 specifically for backward compatibility and make sure we check in indices created by BEST_COMPRESSION in our source tree after every release to make sure we have good bw compatibility coverage.

File: lucene/core/src/test/org/apache/lucene/codecs/lucene80/BaseLucene80DocValuesFormatTestCase.java (hunk 2 above)
Comment (jpountz) [reply lucene/core/src/test/org/apache/lucene/codecs/lucene80/BaseLucene80DocValuesFormatTestCase.java:281]: It's not randomizing, we are testing both modes explicitly via TestBestSpeedLucene80DocValuesFormat on one hand and TestBestCompressionLucene80DocValuesFormat on the other hand.

File: lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/lucene87/Lucene87Codec.java (hunk 1 above)
Comment (jpountz) [reply lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/lucene87/Lucene87Codec.java:71]: Right. It's still possible to made different choices for stored fields and doc values given that we allow configuration of doc values on a per-field basis, but this should at least keep simple use simple with one switch that configures stored fields and doc values at the same time.

Comment (mikemccand) [reply lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/lucene87/Lucene87Codec.java:71]: Great!  Simple for common use cases ("I want best compression" or "I want fastest search"), and complex for complex use cases (I want separate control for each part of the index).

**Review by mikemccand — APPROVED** [review]
Thanks @jpountz!

**Review by mikemccand — APPROVED** [review]
Thanks @jpountz!

**Review by mikemccand — APPROVED** [review]

**Pull request merged**